crates/rpc/src/v03/method/get_events.rs

```diff
@@ -200,97 +176,222 @@ pub async fn get_events(
-                requested_offset.unwrap_or_default() - count
-            }
-            None => 0,
-        };
+    // Append pending data if required.
+    if let Some(pending_block) = pending_block {
+        if matches!(request.to_block, Some(Pending)) {
+            let keys: Vec<std::collections::HashSet<_>> = request
+                .keys
+                .into_iter()
+                .map(|keys| keys.into_iter().collect())
+                .collect();
 
-        let is_last_page = append_pending_events(
-            &pending_block,
-            &mut events.events,
-            skip,
-            amount,
-            request.address,
-            keys,
-        )
-        .await;
-
-        events.continuation_token = next_continuation_token(
-            requested_offset.unwrap_or_default(),
-            request.chunk_size,
-            is_last_page,
-        );
+            if events.events.len() < request.chunk_size {
+                let amount = request.chunk_size - events.events.len();
+
+                let current_offset = match continuation_token {
+                    Some(continuation_token) => {
+                        continuation_token.offset_in_block(pending_block_number)?
+                    }
+                    None => 0,
+                };
+
+                let is_last_page = append_pending_events(
+                    &pending_block,
+                    &mut events.events,
+                    current_offset,
+                    amount,
+                    request.address,
+                    keys,
+                );
+
+                events.continuation_token = if is_last_page {
+                    None
+                } else {
+                    let continuation_token = ContinuationToken {
+                        block_number: pending_block_number,
+                        offset: current_offset + amount,
+                    };
+                    Some(continuation_token.to_string())
+                };
+            } else if page.is_last_page {
+                // the page is full but this was the last page from the DB, so
+                // we should continue with the first pending event in the next
+                // page
+
+                // check if there are matching pending events
+                let mut buf: Vec<types::EmittedEvent> = Vec::new();
+                let _ =
+                    append_pending_events(&pending_block, &mut buf, 0, 1, request.address, keys);
+                if buf.is_empty() {
+                    // if there are no matching events in pending we should not return a token
+                    events.continuation_token = None;
+                } else {
+                    let continuation_token = ContinuationToken {
+                        block_number: pending_block_number,
+                        offset: 0,
+                    };
+                    events.continuation_token = Some(continuation_token.to_string());
+                }
```

Comment:
I'm still on the fence about this.. as in a pending query should never return `None` token. Maybe we should ask.

Reply:
Juno suggests we are fine as is. `None` may be returned.